vendor/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go

```diff
@@ -274,13 +279,24 @@ func (s *GenericAPIServer) PrepareRun() preparedGenericAPIServer {
 // Run spawns the secure http server. It only returns if stopCh is closed
 // or the secure port cannot be listened on initially.
 func (s preparedGenericAPIServer) Run(stopCh <-chan struct{}) error {
+	eventRef := &corev1.ObjectReference{
+		Kind:       "Namespace",
+		Namespace:  "",
+		Name:       eventRefNamespace(),
+		APIVersion: "v1",
+	}
+
 	delayedStopCh := make(chan struct{})
 
 	go func() {
 		defer close(delayedStopCh)
 		<-stopCh
 
+		s.lifeCycleEventRecorder.Eventf(eventRef, corev1.EventTypeNormal, "TerminationStarted", "received signal to terminate, becoming unready, but keeping serving")
+
 		time.Sleep(s.MinimalShutdownDuration)
+
+		s.lifeCycleEventRecorder.Eventf(eventRef, corev1.EventTypeNormal, "MinimalShutdownDurationFinished", "the minimal shutdown duration of %v finished", s.MinimalShutdownDuration)
```

Comment:
i would probably keep the reasons consistent so we can grep? (Termination*)

Reply:
good idea